src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+#include "horizontal_fusion_impl.hpp"
+#include <migraphx/iterator_for.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+static unsigned opcode_bits  = 16;
+static unsigned hash_id_bits = 16;
+static unsigned filter_bits  = 8;
+static unsigned kernel_bits  = 8;
+
+// Register a single operation.
+// 1st arg: operation name. 2nd arg: encoding function.
+void horizontal_fusion_impl::register_op(const std::string& name, encoder func, int flag)
+{
+    op_registry[name] = std::move(func);
+    op_flag[name]     = flag;
+}
+
+// Register operations.
+void horizontal_fusion_impl::register_all()
+{
+    register_op("gpu::convolution", encode_conv_common, 1);
+    register_op("gpu::conv_bias_relu", encode_conv_common, 1);
+    register_op("hip::add_relu", encode_common, 0);
+    register_op("convolution", encode_conv_common, 1);
+    register_op("add", encode_common, 0);
+    register_op("relu", encode_common, 0);
+}
+
+static unsigned opcode_shift_count() { return ((sizeof(key_type) * 8) - opcode_bits); }
+
+static unsigned hash_id_shift_count() { return (opcode_shift_count() - hash_id_bits); }
+
+static unsigned filter_shift_count() { return (hash_id_shift_count() - filter_bits); }
+
+static unsigned kernel_shift_count() { return (filter_shift_count() - kernel_bits); }
+
+// Encode common fields.
+//
+// |----- 16 bits -----|----- 16 bits -------|----- 32 bits -----|
+// |      opcode       | 1st operand hash id |
```

Comment:
This should be written as bit fields rather than comments(as comments can become stale):

```cpp
struct common_encoding
{
    unsigned opcode: 16;
    unsigned operand: 16;
};
```

Reply:
The bits are defined as static constants.  I changed the comment to use names instead.

Reply:
But comments can still be staled. Why can't you use bitfields to define the layout?

Reply:
Why can't I use static constants?